app/routes/api/jobs.py

```diff
@@ -68,60 +72,199 @@ def get_jobs(createdSince=None, createdBefore=None, limit=None):
+    
+    # Get the user...
+    username = body.pop('username')
+    user = db.session.query(User).filter_by(username=username).one_or_none()
+    if user is None:
+        return "User '{}' is not registered".format(username), 403
+    user_id = user.id
+    group_id = user.groups[0].id
+    body['owner'] = user_id
+    body['group'] = group_id
+
+    logger.debug('  user {} is {}:{}'.format(username, user_id, group_id))
+    
     # Get the flowchart and put it in place
+    flowchart_text = body.pop('flowchart')
     flowchart_data = {
-        'path': 'unknown',
-        'json': job_data.pop('flowchart'),
-        'description': 'there is no description!'
+        'id': hashlib.md5(flowchart_text.encode('utf-8')).hexdigest(),
+        'owner': user_id,
+        'group': group_id,
+        'text': flowchart_text,
+        'json': '\n'.join(flowchart_text.splitlines()[2:])
     }
-    flowchart_data['id'] = hashlib.md5(
-        flowchart_data['flowchart_file'].encode('utf-8')
-    ).hexdigest()
-
 
     # Validate and deserialize the flowchart data
     flowchart_schema = FlowchartSchema(many=False)
     try:
-        data = flowchart_schema.load(flowchart_data, session=db.session)
+        flowchart_schema.load(flowchart_data, session=db.session)
     except ValidationError as err:
         return err.messages, 422
 
-    flowchart = Flowchart(**flowchart_data)
-
-    found = db.session.query(Flowchart).filter_by(id=flowchart_data['id']).all()
-    if found:
-        flowchart_id = found[0].id
+    # See if it already in the db, if not, add it
+    flowchart_is_new = False
+    flowchart = db.session.query(Flowchart).filter_by(
+        id=flowchart_data['id']
+    ).one_or_none()
+    if flowchart:
+        if project not in flowchart.projects:
+            flowchart.projects.append(project)
+            db.session.commit()
+        
     else:
+        flowchart = Flowchart(**flowchart_data)
+        flowchart.projects.append(project)
         db.session.add(flowchart)
         db.session.commit()
-        flowchart_id = flowchart.id
+        flowchart_is_new = True
+    flowchart_id = flowchart.id
 
     # Validate and deserialize the job data
-    job_data['status'] = 'Submitted'
-    job_data['flowchart_id'] = flowchart_id
+    body['status'] = 'Submitted'
+    body['flowchart_id'] = flowchart_id
 
     job_schema = JobSchema(many=False)
     try:
-        data = job_schema.load(job_data, session=db.session)
+        job_schema.load(body, session=db.session)
     except ValidationError as err:
         logger.error('ValidationError (job): {}'.format(err.messages))
         logger.info('   valid data: {}'.format(err.valid_data))
         return err.messages, 422
 
+    job = Job(**body)
 
-    job = Job(**job_data)
+    # Get the unique ID for the job...
+    job_id_file = os.path.join(datastore, 'job.id')
+    job.id = get_job_id(job_id_file)
 
+    # And the path
+    project_path = Path(datastore) / 'projects' / project_name
```

Comment:
This could be done with `os.path.join` if we choose to use `os.path`

Reply:
As I noted, I am coming to prefer pathlib. For actual paths they are similar, but for finding permissions on files, expanding home directories and such pathlib hads a clear advantage in my opinion.